### PR TITLE
feat: add support for graph that includes only circular deps

### PR DESCRIPTION
**Pull request by gaspardip**
This is my take on #223 , tested it against a medium size app and it worked fine.

Let me know if there's anything you would like me to change

### CODE REVIEW

**Comment by gaspardip** [comment]
I guess this doesn't address combining `--circular` with other options other than `--image`

**Comment by pahen** [comment]
Sorry, swamped in work right now. I'll take a look at it really soon!

**Comment by gaspardip** [comment]
Sure, no problem! I need to fix some tests here but I'll continue this branch

**Comment by mathieubruguier** [comment]
Any updates on this by any chance? 
Seems super promising and something I could definitely use 👍

File: lib/api.js [hunk 1]
```diff
@@ -152,6 +152,23 @@ class Madge {
...
 
+	circularImage(imagePath) {
+		if (!imagePath) {
+			return Promise.reject(new Error('imagePath not provided'));
+		}
+
+		const circular = this.circular();
+
+		const newTree = Object.entries(this.obj())
+			.filter(([k, v]) => circular.some((x) => x.includes(k)))
```
Comment (RamarajaRamanujan) [inline lib/api.js]:
```suggestion
			.filter(([k]) => circular.some((x) => x.includes(k)))
```
Remove variable ```v``` as it's unused. The build is failing as this is causing the following linting error,
```error  'v' is defined but never used  no-unused-vars```

File: bin/cli.js [hunk 2]
```diff
@@ -202,12 +202,23 @@ new Promise((resolve, reject) => {
...
 
+			if (program.image) {
+				return res.circularImage(program.image).then((imagePath) => {
+					spinner.succeed(
+						`${chalk.bold('Image created at')} ${chalk.cyan.bold(imagePath)}`
+					);
+					return res;
+				});
+			}
+
```
Comment (RamarajaRamanujan) [inline bin/cli.js]: Delete this addition.
```suggestion
```
These lines cause the build to fail with the following linting error,
```
error  Arrow function has a complexity of 11                               complexity
error  Arrow function has too many statements (26). Maximum allowed is 25  max-statements
```
I have a solution for this.
You can add a check for it in the ```image``` method of ```lib/api.js``` at line 151 ,
```
if (this.circular) {
    return this.circularImage(imagePath);
} else {
    return graph.image(this.obj(), this.circular(), imagePath, this.config);
}
```

File: lib/api.js [hunk 3]
```diff
@@ -152,6 +152,23 @@ class Madge {
 		return graph.image(this.obj(), this.circular(), imagePath, this.config);
```
Comment (RamarajaRamanujan) [inline lib/api.js]:
```suggestion
if (this.circular) {
    return this.circularImage(imagePath);
} else {
    return graph.image(this.obj(), this.circular(), imagePath, this.config);
}
```
We can add the check here to generate image for --circular dependencies option.
This is better as it solves the linting issue of complexity and too many statements which is causing the build to fail

File: bin/cli.js [hunk 4]
```diff
@@ -202,12 +202,23 @@ new Promise((resolve, reject) => {
...
 
 		if (program.image) {
 			return res.image(program.image).then((imagePath) => {
-				spinner.succeed(`${chalk.bold('Image created at')} ${chalk.cyan.bold(imagePath)}`);
+				spinner.succeed(
+					`${chalk.bold('Image created at')} ${chalk.cyan.bold(imagePath)}`
+				);
 				return res;
 			});
 		}
```
Comment (RamarajaRamanujan) [inline bin/cli.js]: Move this to condition check to be above ```if (program.circular)```.

**Comment by SaeedZhiany** [comment]
@gaspardip @pahen 

can you please merge and add this feature any soon? 

Thanks

**Comment by pahen** [comment]
Thanks for you contribution @gaspardip. This is now included in v3.10.0 and is published to NPM :)

**Comment by gaspardip** [comment]
awesome!

**Comment by SaeedZhiany** [comment]
@gaspardip 

is it possible to combine --circular with --dot option now?

**Comment by SaeedZhiany** [comment]
@gaspardip 

is it possible to combine `--circular` option with `--dot` option now? if yes how?

I tried to run the both commands `madge --circular --dot .\src\index.tsx > graph.gv` and `madge --circular --dot .\src\index.tsx > graph.dot`. but result is not usage as a `.dot` file in applications like Gephi. because it actually just flush console log into the output file and it's not formatted properly.

**Pull request closed without merging**